File: 0325/SoohyunKim/1149_RGB거리.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const [N, ...cost] = input.map(v => v.split(' ').map(Number));
const dp = cost;

for (let i = 1; i < N; i++){
    dp[i][0] += Math.min(dp[i-1][1], dp[i-1][2]); // i번째 집이 R일 때, 이전 집은 G B중 최소 가격인 것으로 선택
    dp[i][1] += Math.min(dp[i-1][0], dp[i-1][2]); // G
    dp[i][2] += Math.min(dp[i-1][0], dp[i-1][1]); // B
}

console.log(Math.min(...dp[N-1]));

File: 0325/SeoyeonHong/1149 RGB거리.py
import sys

input = sys.stdin.readline
N = int(input())
cost = []
min_cost = [[0, 0, 0] for _ in range(N)]

for _ in range(N):
    cost.append(list(map(int, input().split())))

min_cost[0] = cost[0]

for idx in range(1, N):
    for color in range(3):
        min_cost[idx][color] = cost[idx][color] + min(min_cost[idx-1][(color+1) % 3], min_cost[idx-1][(color+2) % 3])
    
print(min(min_cost[-1]))


# 시간 초과 - 우선순위 큐
# import sys
# import heapq

# input = sys.stdin.readline
# N = int(input())
# cost = []

# for _ in range(N):
#     cost.append(list(map(int, input().split())))

# q = [(0, 0, -1)]
# heapq.heapify(q)

# while q:
#     total_cost, idx, prev_color = heapq.heappop(q)
#     if idx == N:
#         print(total_cost)
#         break
#     for color in range(3):
#         if prev_color != color:
#             heapq.heappush(q, (total_cost + cost[idx][color], idx+1, color))
